### ISSUE TITLE
Multiple performance regressions with ContentCache and FileBackend

### ISSUE THREAD

**Issue by hlubek** [neos/flow-development-collection#1910]
- This seems to be caused by the addition of workspace hashes in cache tags (which themself are a performance improvement)
- Now many more tags seem to be flushed for any node change (e.g. 50 in a recently updated project, but this could be more)
- For each tag, the backend iterates through _all_ cache files (well, there's no index structure yet)
- For every cache entry, there's actually two calls to `readCacheFile` which perform a series of `fopen`, `flock`,  `fread`, `flock` and `fclose`
- This adds up to about 2.000.000 file operations for only ~6200 cache entries (profiled with Tideways)
- This has a dramatic impact on performance
- Given, the FileBackend is not optimal for production deployments, but we can do much better
- I would expect Neos/Flow to read every cache file only once
- Idea: Implement a batched `flushByTags` method and new interface to only iterate once over all files

See https://github.com/neos/flow-development-collection/blob/fdb7b00603b3982025047bd84ad1b115bfe566a9/Neos.Cache/Classes/Backend/FileBackend.php#L291 and https://github.com/neos/flow-development-collection/blob/fdb7b00603b3982025047bd84ad1b115bfe566a9/Neos.Cache/Classes/Backend/SimpleFileBackend.php#L495

**Comment by kitsunet** [comment]
Yep, as I feared :( 

Adding `flushByTags` would mean in interface change though....

**Comment by hlubek** [comment]
> Yep, as I feared :(
> 
> Adding `flushByTags` would mean in interface change though....

Well, we could add something like a `BatchedTagFlushBackendInterface` with only this method and check for it when calling `flushByTags` in `AbstractFrontend` (which also needs an additional interface for compatibility) - otherwise we could fall back to iterating through tags and calling `flushByTag`. So it could be compatible with existing backends and frontends.

**Comment by kitsunet** [comment]
Right, that could probably work! I like

**Comment by albe** [comment]
Yep, sounds good to me too. Regarding the 2 `readCacheFile` calls - we store the position of the metadata at the end of the file which is needed to find the start of it, so the only solution to avoid two calls would be to:
- create a specific method for `readCacheMetadata` that only does one `fopen`, `fclose`, `flock` but still two `fread`s - duplicates most of the `readCacheFile`
- read the whole file in one go and parse out the metadata in-memory - can be permissive with relatively large cache files

**Comment by hlubek** [comment]
>     * create a specific method for `readCacheMetadata` that only does one `fopen`, `fclose`, `flock` but still two `fread`s - duplicates most of the `readCacheFile`

Yes, that's what I would prefer. Maybe we should create a private function that takes another callable to perform the actual code with locking (e.g. `withLock`). This way the locking is only implemented once.

**Comment by albe** [comment]
I guess that would be fine with me :)
